File: python/102/patterns/singleton1.py
class Singleton:
    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            instance = super().__new__(cls, *args, **kwargs)
            cls._instance = instance
        return cls._instance


s1 = Singleton()
s2 = Singleton()
print(id(s1), id(s2))

File: c#-basics/patterns/singleton1.cs
namespace ConsoleApp18
{
    public class Singleton
    {

        private static Singleton _instance;

        private Singleton()
        {
            Console.WriteLine("Singleton instance created.");
        }
        
        public static Singleton Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = new Singleton();
                }
                return _instance;
            }
        }
        
        public void DoSomething()
        {
            Console.WriteLine("Singleton is doing something!");
        }
    }
    class Program
    {
        static void Main(string[] args)
        {
            
            Singleton s1 = Singleton.Instance;
            s1.DoSomething();
           
            Singleton s2 = Singleton.Instance;
           
            if (s1 == s2)
            {
                Console.WriteLine("Both instances are the same!");
            }
        }
    }
}
